dlls/ff/ff_player.cpp

```diff
@@ -8009,3 +8030,135 @@ void CFFPlayer::UpdateCamera( bool bUnassigned )
 		}				
 	}
 }
+
+// purpose: update our vector of recent attackers for kill assists
+void CFFPlayer::AddRecentAttacker( const CTakeDamageInfo &dmgInfo )
+{
+	// make sure to get the effective scorer incase dmg is from
+	// a buildable - det, sg, dispenser, probably nades etc	
+	CFFPlayer *pAttacker = ToFFPlayer( FFGameRules()->GetDeathScorer( dmgInfo.GetAttacker(), dmgInfo.GetInflictor() ) );
+
+	// dont track our own client in recent attacks, otherwise we would show up as an assist when the world kills us & suicides
+	// note: world dmg comes in as null here.
+	if ( !pAttacker || pAttacker == this )
+		return;
+
+	int attackerIdx = pAttacker->entindex( );
+	float dmg = dmgInfo.GetDamage( );
+	float timestamp = gpGlobals->curtime;
+
+	DevMsg( "CFFPlayer::AddRecentAttacker0: index = %d dmg = %f timestamp = %f\n", attackerIdx, dmg, timestamp );
+	// search for existing or create new
+	for ( int i = 0; i < m_recentAttackers.Count( ); ++i )
+	{
+		if ( m_recentAttackers[i].playerIndex == attackerIdx || pAttacker == m_recentAttackers[i].pFFPlayer )
+		{
+			m_recentAttackers[i].totalDamage += dmg;
+			m_recentAttackers[i].timestamp = timestamp;
+			DevMsg( "CFFPlayer::AddRecentAttacker1: totalDamage = %f\n", m_recentAttackers[i].totalDamage );
+			return;
+		}
+	}
+
+	// if we didnt find a match, create & add new
+	DevMsg( "CFFPlayer::AddRecentAttacker2\n" );
+	//m_recentAttackers.AddToHead( RecentAttackerInfo( attackerIdx, dmg, timestamp, pAttacker ) );
+	m_recentAttackers.AddToTail( RecentAttackerInfo( attackerIdx, dmg, timestamp, pAttacker ) );
+}
+
+// returns assisted attacker that did most damage within the last MAX_ASSIST_TIME_MS or NULL if nothing found
+RecentAttackerInfo* CFFPlayer::GetTopKillAssister( CBasePlayer* killerToIgnore )
+{
+	// oops: world kills will come in null
+	bool killedByWorld = killerToIgnore == NULL;
+	const char* killerName = killedByWorld ? "world" : killerToIgnore->GetPlayerName( );
+	DevMsg( "CFFPlayer::GetTopKillAssister0: %s killer='%s'\n", GetPlayerName( ), killerName );
+
+	RecentAttackerInfo* ret = NULL;
+
+	for ( int i = 0; i < m_recentAttackers.Count( ); ++i )
+	{
+		CFFPlayer* pFFAssister = m_recentAttackers[i].pFFPlayer;
+		// this shouldnt happen, but you know
+		if ( pFFAssister == NULL )
+			continue;
+
+		// if this was a suicide and we are killerToIgnore, we dont have to check here
+		// because previous logic prevents adding ourselves to the assist list
+		bool isKiller = !killedByWorld && pFFAssister == killerToIgnore;
+
+#ifdef _DEBUG
+		DevMsg( "CFFPlayer::GetTopKillAssister1: assister='%s' cum dmg='%f' isKiller='%s'\n",
+			pFFAssister->GetPlayerName( ), m_recentAttackers[i].totalDamage, isKiller ? "yes" : "no" );
+#endif

```

Comment:
If all this works then I would remove the dev messages, just clutters the code and the console for the next guy